artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/MultiIterator.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.utils.collections;
+
+import java.util.Iterator;
+
+/**
+ * Provides an Iterator that works over multiple underlying iterators.
+ *
+ * @param <T> type of the class of the iterator.
+ */
+public class MultiIterator<T> implements Iterator<T> {
```

Comment:
Yeah i had a version like that but it makes using simple MultiIterator ugly to use. Uneededly. And to extend if needed only needs a cast. In case of resetable.

Reply:
OK, so to do this and keep the cleaner user interface what i can do is make an abstract base which has most of this logic in, and have the more ugly generics. And then make MultiIterator extend that which then keeps its cleaner generics, and like wise MultiResettableIterator. Ill update PR to see if this is good compromise, or if you'd prefer me to revert and simply keep the cast i had originally.